src/helpers/menu_service.ts

```diff
@@ -0,0 +1,70 @@
+import { MenuProps } from "../components/menu/menu";
+import { UuidGenerator } from "./uuid";
+
+interface RegisterMenuArgs extends Omit<MenuProps, "onClose" | "depth"> {
+  onClose?: () => void;
+}
+
+interface InternalMenu extends RegisterMenuArgs {
+  id: string;
+}
+
+export class MenuService {
+  private uuidGenerator = new UuidGenerator();
+
+  private activeMenu: InternalMenu | undefined;
+
+  constructor() {}
+
+  registerMenu(menu: RegisterMenuArgs) {
+    if (this.activeMenu) {
+      this.activeMenu.onClose?.();
+    }
+
+    const id = this.uuidGenerator.uuidv4();
+    const newMenu = {
+      ...menu,
+      id,
+    };
+    this.activeMenu = newMenu;
+
+    return newMenu.id;
+  }
+
+  getCurrentMenuId() {
+    return this.activeMenu?.id;
+  }
+
+  getCurrentMenuProps(): MenuProps | undefined {
+    if (!this.activeMenu) return undefined;
+    const menu = this.activeMenu;
+    const props: MenuProps = {
+      ...menu,
+      depth: 1,
+      onClose: () => {
+        if (this.activeMenu?.id === menu.id) this.activeMenu = undefined;
+        menu.onClose?.();
+      },
+    };
+    delete props["id"];
+    return props;
+  }
+
+  hasOpenMenu(): boolean {
+    return !!this.activeMenu;
+  }
+
+  unregisterMenu(id: string | undefined) {
+    if (id && this.activeMenu?.id === id) {
+      // call onClose after setting activeMenu to undefined to avoid possible infinite recursion
```

Comment:
The problem occurs if you call `mernuService.unregisterMenu()` from the callback `onClose` of the menu. I feel like it's a very easy mistake to do, and preventing it from making everything crash doesn't cost much.